### ISSUE TITLE
Issue with docker persistence

### ISSUE THREAD

**Issue by NicolasCARPi** [robiso/wondercms#92]
So while playing a bit with the software, I realized that because the root of the project is the mounted directory as volume for persistence, updating the image doesn't actually do anything as the whole `/var/www/html` folder is mounted. Containers should be stateless, so one should be able to update the image and launch a new container with fresh code and keep data.

My suggestion would be to separate the actual data from the code. Create a `data` folder containing the database and the files (everything user generated let's say). And this data folder can then be mounted as a volume in docker, without having to mount the whole project with the code (code !== userdata).

This would be a step in the right direction imho. Also for backups it would be better because one can only backup the `data` folder, not the whole project. Only user data is important, code from project can be obtained from github.

Once this is done I'll have a closer look at the `.htaccess-ultimate` and make it default in the docker container (or at least part of it). Some security headers should be there for sure.

During update, the code would need to check the existence of the `data` folder, create it and move the database in there and the files. But do you have an update function able to do this stuff between release?

**Comment by robiso** [comment]
@NicolasCARPi 

The only user generated file is in the database.js, which contains all of the data.
The backup feature makes sure any changes aren't lost to the users custom theme. Migrating the users site is also easier with the "backup zip" and uploading it to the new host without dealing with data and theme changes separately.

Here's an example of how we handled updating the database and database migrations in the past.
```
public static function _updateOtherFiles() {
    $olddb = wCMS::db();
    if ( ! isset($olddb->{'config'}->{'dbVersion'})) {
        if (file_exists(__DIR__ . '/themes/default/theme.php')) file_put_contents(__DIR__ . '/themes/default/theme.php', wCMS::_getExternalFile('https://raw.githubusercontent.com/robiso/wondercms/master/themes/default/theme.php'));
        if (file_exists(__DIR__ . '/themes/default/css/style.css')) file_put_contents(__DIR__ . '/themes/default/css/style.css', wCMS::_getExternalFile('https://raw.githubusercontent.com/robiso/wondercms/master/themes/default/css/style.css'));
        if (file_exists('.htaccess')) file_put_contents('.htaccess', wCMS::_getExternalFile('https://raw.githubusercontent.com/robiso/wondercms/master/.htaccess'));
        unlink('database.js');
        wCMS::_createDatabase();
        $newdb = wCMS::db();
        $newdb->{'config'}->{'siteTitle'} = $olddb->{'config'}->{'siteTitle'};
        $newdb->{'config'}->{'theme'} = 'default';
        $newdb->{'config'}->{'defaultPage'} = $olddb->{'config'}->{'defaultPage'};
        $newdb->{'config'}->{'login'} = $olddb->{'config'}->{'login'};
        $newdb->{'config'}->{'password'} = $olddb->{'config'}->{'password'};
        $newdb->{'config'}->{'menuItems'} = $olddb->{'config'}->{'menuItems'};
        $newdb->{'pages'} = $olddb->{'pages'};
        $newdb->{'blocks'}->{'subside'}->{'content'} = $olddb->{'config'}->{'subside'};
        $newdb->{'blocks'}->{'footer'}->{'content'} = $olddb->{'config'}->{'copyright'};
        wCMS::save($newdb);
    }
    if ($olddb->config->dbVersion < '2.1.0') {
        $newMenu = new stdClass();
        $newPages = new stdClass();
        $i = 0;
        foreach ($olddb->pages as $pageName => $page) {
            $slugPage = wCMS::_slugify($pageName);
            $newPages->$slugPage = $page;
            $newMenu->{$i} = new stdClass;
            $newMenu->{$i}->name = $page->title;
            $newMenu->{$i}->slug = $slugPage;
            $newMenu->{$i}->visibility = "show";
            $i++;
        }
        unset($olddb->config->menuItems);
        unset($olddb->pages);
        $olddb->config->menuItems = $newMenu;
        $olddb->pages = $newPages;
        $olddb->config->dbVersion = '2.1.0';
        wCMS::save($olddb);
    } elseif ($olddb->config->dbVersion < '2.3.0') {
        if (file_exists(__DIR__ . '/themes/default/theme.php')) file_put_contents(__DIR__ . '/themes/default/theme.php', wCMS::_getExternalFile('https://raw.githubusercontent.com/robiso/wondercms/master/themes/default/theme.php'));
        if (file_exists(__DIR__ . '/themes/default/css/style.css')) file_put_contents(__DIR__ . '/themes/default/css/style.css', wCMS::_getExternalFile('https://raw.githubusercontent.com/robiso/wondercms/master/themes/default/css/style.css'));
        $olddb->config->dbVersion = '2.3.0';
        wCMS::save($olddb);
    }
}
```

In my opinion, WonderCMS has more novice users than any other CMS (proportional to user base) and wouldn't know how to deal with backing up the data separately, and then copying any changes made in their theme. More advanced users probably choose Grav as their top pick and I'm completely fine with that. You can't satisfy everyone and I personally find Grav harder to use but still very much love what it stands for. :)

Is there a possible workaround for having a data folder and backing everything up?

**Comment by NicolasCARPi** [comment]
> Is there a possible workaround for having a data folder and backing everything up?

Yes, I think it's not an issue to have a separate data folder but still keep backing up the whole project.


> only user generated file is in the database.js

Isn't there also the `files` folder with the uploaded files?

**Comment by robiso** [comment]
@NicolasCARPi 
> Isn't there also the files folder with the uploaded files?

Nice catch, no idea how I missed that 🤦‍♂️, guess it's too early! There's also the plugins folder where only the contact form comes to mind, which currently has to be edited manually with the users email address to work properly. There's a improved version of the contact form plugin, which keeps a "config" file, so the user doesn't have to update/overwrite other files when updating (https://github.com/Jeremy1980/WonderCMS-contactForm).

**Comment by NicolasCARPi** [comment]
Fixed by #113
